models/baseModels/tests/testLoyaltyProgram.spec.ts

```diff
@@ -0,0 +1,275 @@
+    return loyaltyPointEntryData[0];
+  }
+}
+
+async function createSalesInvoice() {
+  const sinvDoc = fyo.doc.getNewDoc(ModelNameEnum.SalesInvoice, {
+    account: 'Debtors',
+    party: partyData.name,
+    items: [
+      {
+        item: itemData.name,
+        rate: itemData.rate,
+        quantity: 1,
+      },
+    ],
+  }) as SalesInvoice;
+
+  return sinvDoc;
+}
+
+test('create Sales Invoice and verify loyalty points are created correctly', async (t) => {
+  const sinvDoc = await createSalesInvoice();
+
+  await sinvDoc.sync();
+  await sinvDoc.runFormulas();
+  await sinvDoc.submit();
+
+  t.ok(
+    fyo.db.exists(ModelNameEnum.SalesInvoice, sinvDoc.name),
+    `Sales Invoice '${sinvDoc.name}' exists`
+  );
+
+  t.ok(
+    fyo.db.exists(ModelNameEnum.SalesInvoice, sinvDoc.loyaltyProgram),
+    `Loyalty Program '${sinvDoc.loyaltyProgram}' should be linked to Sales Invoice '${sinvDoc.name}'`
+  );
+
+  const loyaltyPointEntryData = await loyaltyPointEntryDoc(
+    sinvDoc.name as string
+  );
+
+  t.ok(
+    await fyo.db.exists(
+      ModelNameEnum.LoyaltyPointEntry,
+      loyaltyPointEntryData?.name
+    ),
+    `Loyalty Point Entry '${loyaltyPointEntryData?.name}' should be created against Sales Invoice '${sinvDoc.name}'`
+  );
+
+  const loyaltyProgramDoc = (await fyo.doc.getDoc(
+    ModelNameEnum.LoyaltyProgram,
+    sinvDoc.loyaltyProgram
+  )) as Party;
+
+  const selectedTier: CollectionRulesItems | undefined = getLoyaltyProgramTier(loyaltyProgramDoc, fyo.pesa(itemData.rate));
+
+  t.equals(
+    loyaltyPointEntryData?.loyaltyProgramTier,
+    selectedTier?.tierName,
+    `tier must be ${loyaltyPointEntryData?.loyaltyProgramTier} because it is the matched tier'`
+  );
+
+  const tierData: any = collectionRulesData.find((rule) => {
+    return rule.tierName === loyaltyPointEntryData?.loyaltyProgramTier;
+  });
+
+  t.equals(
+    loyaltyPointEntryData?.loyaltyPoints,
+    (itemData.rate * tierData?.collectionFactor) as number,
+    `${loyaltyPointEntryData?.loyaltyPoints} loyalty points should be added to customer '${loyaltyPointEntryData?.customer}'`
+  );
+});
+
+test('create Sales Invoice with future date and verify loyalty points are not created', async (t) => {
+  const futureDate = new Date(new Date().setDate(new Date().getDate() + 20));
+
+  const sinvDoc = await createSalesInvoice();
+  sinvDoc.date = futureDate;
+
+  await sinvDoc.sync();
+  await sinvDoc.runFormulas();
+  await sinvDoc.submit();
+
+  t.ok(
+    fyo.db.exists(ModelNameEnum.SalesInvoice, sinvDoc.name),
+    `Sales Invoice '${sinvDoc.name}' exists`
+  );
+
+  const loyaltyPointEntryData = await loyaltyPointEntryDoc(
+    sinvDoc.name as string
+  );
+
+  t.equals(
+    loyaltyPointEntryData?.name,
+    undefined,
+    'Loyalty points should not be created for a future-dated Sales Invoice'
+  );
+});
+
+test('create Sales Invoice with Loyalty Points redemption and verify negative loyalty point entry is created correctly', async (t) => {
```

Comment:
can be simplified to something like,
_redeem loyalty points and verify a new loyalty point entry doc is created_